### PR TITLE
Add new simpler and more explicit syntax for check-cfg

### LINKED ISSUE #636: Simplify and improve explicitness of the check-cfg syntax
# Proposal

This proposal is a spawned of work started in [this PR](https://github.com/rust-lang/rust/pull/111072). Current documentation for check-cfg can be found [in the unstable book](https://doc.rust-lang.org/unstable-book/compiler-flags/check-cfg.html).

<details>
<summary>
Old proposal
</summary>

This proposal wants to adds a new simpler and more explicit syntax for check-cfg. It would consist of two new form [[detailed doc](https://github.com/rust-lang/rust/blob/6846f334779ebe5de1d902e7e22cb145c2979ce4/src/doc/unstable-book/src/compiler-flags/check-cfg.md#check-cfg)]:

1. `--check-cfg exhaustive(...)` enables some exhaustive checking
2. `--check-cfg configure(...)` enables checking the values within list-valued conditions.

## The `exhaustive(...)` form

The `exhaustive(...)` form tell the compiler if it should be exhaustive about it's checking.

It currently has two exhaustiveness:
 - `names`: permist the compiler to check every config name by adding it's list of well-known names
 - `values`: permist the compiler to check it's well-known cfg values by adding it's list of well-known values

## The `configure(...)` form

The `configure(...)` form enables checking the values within list-valued conditions. It has this form:

```bash
rustc --check-cfg `configure(name1, ..., nameN, "value1", "value2", ... "valueN")'
```

To enable checking of values, but to provide an empty set of valid values, use this form:

```bash
rustc --check-cfg `configure(name1, ..., nameN)`
```

The `--check-cfg configure(...)` option can be repeated, both for the same condition name and for different names. If it is repeated for the same condition name, then the sets of values for that condition are merged together.

## Why?

The preview forms `names(...)` and `values(...)` have implicit meaning that are not strait-forward. In particular `values(foo)`&`values(bar)` and `names(foo, bar)` are not equivalent which has created [some confusions](https://github.com/rust-lang/rust/pull/98080).

Also the `names()` and `values()` form are not clear either and again created some confusions where peoples believed that `values()`&`values(foo)` could be reduced to just `values(foo)`.

To fix that the two new forms are made to be explicit and simpler. See the table of correspondence:
  - `names()` -> `exhaustive(names)`
  - `values()` -> `exhaustive(values)`
  - `names(foo)` -> `exhaustive(names)`&`configure(foo)`
  - `values(foo)` -> `configure(foo)`
  - `values(feat, "foo", "bar")` -> `configure(feat, "foo", "bar")`
  - `values(foo)`&`values(bar)` -> `configure(foo, bar)`
  - `names()`&`values()`&`values(my_cfg)` -> `exhaustive(names, values)`&`configure(my_cfg)`

*The two previous forms would be deprecated and will would be removed once cargo and beta rustc have the necessary support.*

## Open questions

 - Naming of `configure` ([comment](https://github.com/rust-lang/rust/pull/111072#issuecomment-1544023462)):
   > `--check-cfg configure(...)` also sounds tautological, `cfg` already means "configure", `values` was a better name for this semantic, IMO.
   
    Could be kept as `values(...)` without too much difficulties, having a different name seemed clearer that it was different.
 - Naming of `exhaustive`:
   Too different from `names` to re-use the same name.

</details>

This proposal is based on the previous proposal above and the summary work done by @wesleywiser in https://hackmd.io/@wesleywiser/rJlumbfep.

This MCP proposes to:
 - remove the `names()` and `values()` form of `--check-cfg` (with at least one release of deprecation)
 - replace those two forms by new unified form one called `cfg()` (to remove most of un-intuitiveness of the previous syntax)

## The `cfg(...)` form

The `cfg(...)` form enables checking the values within list-valued conditions. It has this form:

```bash
rustc --check-cfg 'cfg(name1, ..., nameN, values("value1", "value2", ... "valueN"))'
```

To enable checking of values, but to provide an empty set of expected values, use these forms:

```bash
rustc --check-cfg 'cfg(name1, ..., nameN)'
rustc --check-cfg 'cfg(name1, ..., nameN, values())'
```

To enable checking of name but not values (i.e. unknown expected values), use this form:

```bash
rustc --check-cfg 'cfg(name1, ..., nameN, values(any()))'
```

To disable checking of name, use this form:

```bash
rustc --check-cfg 'cfg(any())'
```

The `--check-cfg cfg(...)` option can be repeated, both for the same condition name and for different names. If it is repeated for the same condition name, then the sets of values for that condition are merged together.

## Built-in names and values

 - Built-in values checking would always be activated as long as a `--check-cfg` argument is present 
 - Built-in names checking would always be activated as long as a `--check-cfg` argument is present **unless** if any `cfg(any())` arg is passed
 - If one want to enable values and names checking without having any cfg to declare, one can use an empty `cfg()`

## Why?

The preview forms `names(...)` and `values(...)` have implicit meaning that are not strait-forward. In particular `values(foo)`&`values(bar)` and `names(foo, bar)` are not equivalent which has created [some confusions](https://github.com/rust-lang/rust/pull/98080).

Also the `names()` and `values()` form are not clear either and again created some confusions where peoples believed that `values()`&`values(foo)` could be reduced to just `values(foo)`.

To fix that the two new forms are made to be explicit and simpler. See the table of correspondence:
  - `names()` -> enable-by-default by any use of `--check-cfg` useless given `cfg(any())`
  - `values()` -> enable-by-default by any use of `--check-cfg`
  - `names(foo)` -> `cfg(foo, any())`
  - `values(foo)` -> `cfg(foo)`
  - `values(feat, "foo", "bar")` -> `cfg(feat, values("foo", "bar"))`
  - `values(foo)`&`values(bar)` -> `cfg(foo, bar)`
  - `names()`&`values()`&`values(my_cfg)` -> `cfg(my_cfg)`

*The two previous forms would be deprecated and will would be removed once cargo and beta rustc have the necessary support.*

## Open questions

 - Naming of `cfg` form ([comment](https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/Simplify.20and.20improve.20explicitness.20of.20the.20.E2.80.A6.20compiler-team.23636/near/393744633)):
   > `--check=cfg(...)`. I had this thought as well but was struggling to think of other things it could be extended to. It also seemed like users might think `rustc --check ...` is equivalent to `cargo check`. Need to think about this more...
   
   `rustdoc` already has a `--check` argument making `--check=cfg(...)` impossible

# Mentors or Reviewers

@petrochenkov for reviewing and my-self for the implementation work

# Process

The main points of the [Major Change Process][MCP] are as follows:

* [x] File an issue describing the proposal.
* [ ] A compiler team member or contributor who is knowledgeable in the area can **second** by writing `@rustbot second`.
    * Finding a "second" suffices for internal changes. If however, you are proposing a new public-facing feature, such as a `-C flag`, then full team check-off is required.
    * Compiler team members can initiate a check-off via `@rfcbot fcp merge` on either the MCP or the PR.
* [ ] Once an MCP is seconded, the Final Comment Period begins. If no objections are raised after 10 days, the MCP is considered **approved**.

You can read [more about Major Change Proposals on forge][MCP].

[MCP]: https://forge.rust-lang.org/compiler/mcp.html

# Comments

**This issue is not meant to be used for technical discussion. There is a Zulip stream for that. Use this issue to leave procedural comments, such as volunteering to review, indicating that you second the proposal (or third, etc), or raising a concern that you would like to be addressed.**

### PR DESCRIPTION
**Pull request by Urgau**
<details>
<summary>
Old proposition (before the MCP)
</summary>

This PR adds a new simpler and more explicit syntax for check-cfg. It consist of two new form:
 - `exhaustive(names, values)`
 - `configure(name, "value1", "value2", ... "valueN")`

The preview forms `names(...)` and `values(...)` have implicit meaning that are not strait-forward. In particular `values(foo)`&`values(bar)` and `names(foo, bar)` are not equivalent which has created [some confusions](https://github.com/rust-lang/rust/pull/98080).

Also the `names()` and `values()` form are not clear either and again created some confusions where peoples believed that `values()`&`values(foo)` could be reduced to just `values(foo)`.

To fix that the two new forms are made to be explicit and simpler. See the table of correspondence:
  - `names()` -> `exhaustive(names)`
  - `values()` -> `exhaustive(values)`
  - `names(foo)` -> `exhaustive(names)`&`configure(foo)`
  - `values(foo)` -> `configure(foo)`
  - `values(feat, "foo", "bar")` -> `configure(feat, "foo", "bar")`
  - `values(foo)`&`values(bar)` -> `configure(foo, bar)`
  - `names()`&`values()`&`values(my_cfg)` -> `exhaustive(names, values)`&`configure(my_cfg)`

Another benefits of the new syntax is that it allow for further options (like conditional checking for --cfg, currently always on) without syntax change.

The two previous forms are deprecated and will be removed once cargo and beta rustc have the necessary support.

</details>

This PR is the first part of the implementation of [MCP636 - Simplify and improve explicitness of the check-cfg syntax](https://github.com/rust-lang/compiler-team/issues/636).

## New `cfg` form

It introduces the new [`cfg` form](https://github.com/rust-lang/compiler-team/issues/636) and deprecate the other two:
```
rustc --check-cfg 'cfg(name1, ..., nameN, values("value1", "value2", ... "valueN"))'
```

## Default built-in names and values

It also changes the default for the built-in names and values checking.

 - Built-in values checking would always be activated as long as a `--check-cfg` argument is present 
 - Built-in names checking would always be activated as long as a `--check-cfg` argument is present **unless** if any `cfg(any())` arg is passed

~~**Note: depends on https://github.com/rust-lang/rust/pull/111068 but is reviewable (last two commits)!**~~

Resolve https://github.com/rust-lang/compiler-team/issues/636

r? @petrochenkov

### CODE REVIEW

**Comment by petrochenkov** [comment]
Blocked on https://github.com/rust-lang/rust/pull/111068.

**Comment by petrochenkov** [comment]
How can `exhaustive(values)` be a single flag?
Isn't value checking enabled and disabled per-name?
I.e. we can check values for name `foo`, and not check values for name `bar`.
Or does it only affect cfgs built into the compiler?

**Comment by petrochenkov** [comment]
Does `--check-cfg exhaustive(names, values)` basically mean `--check-cfg builtin(names, values)`?
What does `--check-cfg exhaustive(values)` mean without names?

**Comment by petrochenkov** [comment]
With this new design we cannot specify a name as known, while keeping its list of values unknown, right?

Previously `names(foo)` didn't restrict the set of values for `foo`, but `exhaustive(names) & configure(foo)` does restrict it to "no values".

**Comment by Urgau** [comment]
> How can `exhaustive(values)` be a single flag? Isn't value checking enabled and disabled per-name? I.e. we can check values for name `foo`, and not check values for name `bar`. Or does it only affect cfgs built into the compiler?

`exhaustive(values)` works in the same way as `values()`, it adds (nearly) all the builtin cfgs with their expected values into the expected list. That means that all the builtin cfgs will lint if there is an unexpected value and that's it.
It doesn't change anything for non-builtin cfgs.

Remember that we can only lint on values for which we have an entry for, if we don't have one then it's not a value problem but a name problem and this is gated under `exhaustive(names)`.

> Does `--check-cfg exhaustive(names, values)` basically mean `--check-cfg builtin(names, values)`?

Not exactly, for `values` yes, it's an append of the builtin values. But for `names` it's not only appending the list of builtin names but also enabling the checking of all names; otherwise we wouldn't be able to check names since we would always trigger on builtin names.

> What does `--check-cfg exhaustive(values)` mean without names?

Enable checking of the builtin values without triggering on potentially unexpected cfg.

**Comment by Urgau** [comment]
> With this new design we cannot specify a name as known, while keeping its list of values unknown, right?
> 
> Previously `names(foo)` didn't restrict the set of values for `foo`, but `exhaustive(names) & configure(foo)` does restrict it to "no values".

Yes, this is the only "regression" of the new design, but I've yet to found a example where this would be useful. The only place where it is useful is in the compiler for `feature` and `target_feature` where getting the expected values is non trivial.

If a consumer of `rustc` doesn't know the expected values of a config, how would a actual user know which one to use? Or even how does the consumer of `rustc` know which one to set via `--cfg`?

@rustbot ready

**Comment by petrochenkov** [comment]
So what we really need to cover all cases without regressions (without any syntax beautifications) is:
- `--check-cfg builtin-values` - will be commonly used if we expect full checking to be typically used.
- `--check-cfg custom-values(noval, oneval, "val1", twovals, "val1", "val2")` - will be commonly used if we expect full checking to be typically used.
- `--check-cfg exhaustive-names(foo, bar)` - doesn't put value restrictions on `foo` and `bar`, `--check-cfg exhaustive-names()` without arguments will be commonly used if we expect full checking to be typically used, the form with arguments - not so commonly.

**Comment by petrochenkov** [comment]
In that case we probably need some short syntax for the commonly used combination `--check-cfg builtin-values --check-cfg exhaustive-names() --check-cfg custom-values(noval, oneval, "val1", twovals, "val1", "val2")` + some additional rarely used options for opting out of exhaustiveness and adding names without value requirements.

**Comment by Urgau** [comment]
> So what we really need to cover all cases without regressions (without any syntax beautifications) is:
> 
>  * `--check-cfg builtin-values` - will be commonly used if we expect full checking to be typically used.
> 
>  * `--check-cfg custom-values(noval, oneval, "val1", twovals, "val1", "val2")` - will be commonly used if we expect full checking to be typically used.

Agree for these two.

>  * `--check-cfg exhaustive-names(foo, bar)` - doesn't put value restrictions on `foo` and `bar`,

As I said, I don't think this is useful for users, as users knows all the possible values. Otherwise they couldn't possible use cfgs in any meaningful way.

> `--check-cfg exhaustive-names()` without arguments will be commonly used if we expect full checking to be typically used, the form with arguments - not so commonly.

:+1:  (but note that there isn't currently any plan to enable this by default in cargo)

> In that case we probably need some short syntax for the commonly used combination `--check-cfg builtin-values --check-cfg exhaustive-names() --check-cfg custom-values(noval, oneval, "val1", twovals, "val1", "val2")` + some additional rarely used options for opting out of exhaustiveness and adding names without value requirements.

If cargo were to enable all things, it would be with my syntax `exhaustive(names, values)`&`configure(feature, ...)` set by cargo + users defined (but cannot really do anything about them).

**Comment by petrochenkov** [comment]
`--check-cfg configure(...)` also sounds tautological, `cfg` already means "configure", `values` was a better name for this semantic, IMO.
We could also add some sigil to enable checking built-in values in the same list, like
```
--check-cfg values(*, noval, oneval, "val1", twovals, "val1", "val2")
```
or similar.

`--check-cfg names(...)` could then also stay as is.

**Comment by petrochenkov** [comment]
Let's collect more feedback from people.

**Comment by Urgau** [comment]
> `--check-cfg configure(...)` also sounds tautological, `cfg` already means "configure", `values` was a better name for this semantic, IMO.

Naming is hard and I wanted to avoid re-using the current syntax since it's so confusing.
When declaring an expected cfg with zero values there is currently 2 ways:
 - `names(NAME)`: which declares it with *any* values possibles (which technically means that it accept zero values but it also accept any other values which isn't what the user probably wanted) + it implicitly enables the checking of all names
 - `values(NAME)`: which does the right thing but is weird since `values` would as first glance imply that it's use is for when you have some values not none 

> We could also add some sigil to enable checking built-in values in the same list, like
> 
> ```
> --check-cfg values(*, noval, oneval, "val1", twovals, "val1", "val2")
> ```
> 
> or similar.

`*` is not a valid syntax for attribute (since that's what `--check-cfg` uses for parsing): *expected unsuffixed literal or identifier, found `*`*.
Maybe `values(any())` ?

> `--check-cfg names(...)` could then also stay as is.

I really don't like the implicitness of `names(...)` it makes very awkward to use. Users never expected it's current meaning: add the names with *any* values possibles + enable checking of all names.
Users generally just want to say: here is a expected cfg with zero values and for that they should use `values(NAME)`.

**Comment by Urgau** [comment]
If we want to retain the ability to express the name with the *any* values case (which I'm not sure we should), I could think of this `configure(NAME, any())`.

And maybe instead of `configure`, `declare` might be a better name ?

**Comment by apiraino** [comment]
Complementing the [previous comment](https://github.com/rust-lang/rust/pull/111072#issuecomment-1553154064) with the [Zulip notes](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-05-18/near/359409980) from the meeting

@rustbot label -I-compiler-nominated

**Comment by Urgau** [comment]
I've updated the PR to reflect [MCP 636 - Simplify and improve explicitness of the check-cfg syntax](https://github.com/rust-lang/compiler-team/issues/636).

*This is still waiting for the MCP to be accepted (should be next week), but in the meantime I don't see anything preventing reviews.*
@rustbot label +S-waiting-on-review

File: tests/ui/check-cfg/well-known-names.stderr [hunk 1]
```diff
@@ -2,11 +2,14 @@ warning: unexpected `cfg` condition name
   --> $DIR/well-known-names.rs:6:7
    |
 LL | #[cfg(target_oz = "linux")]
-   |       ---------^^^^^^^^^^
-   |       |
-   |       help: there is a config with a similar name: `target_os`
+   |       ^^^^^^^^^^^^^^^^^^^
    |
+   = help: expected values for `target_os` are: `aix`, `android`, `cuda`, `dragonfly`, `emscripten`, `espidf`, `freebsd`, `fuchsia`, `haiku`, `hermit`, `horizon`, `hurd`, `illumos`, `ios`, `l4re`, `linux`, `macos`, `netbsd`, `none`, `nto`, `openbsd`, `psp`, `redox`, `solaris`, `solid_asp3`, `teeos`, `tvos`, `uefi`, `unknown`, `vita`, `vxworks`, `wasi`, `watchos`, `windows`, `xous`
```
[the remark this answers is not in the document]
Comment (Urgau) [reply tests/ui/check-cfg/well-known-names.stderr]: It was printed because the diagnostics were trying to be helpful, and in this case were trying a bit too hard. Pushed a fix as a separate commit (since it's an pre-existing issue).

File: tests/ui/check-cfg/unknown-values.with_values.stderr [hunk 2]
```diff
@@ -0,0 +1,7 @@
+warning: unexpected `with_values` as condition name
```
[the remark this answers is not in the document]
Comment (Urgau) [reply tests/ui/check-cfg/unknown-values.with_values.stderr]: > Aren't warnings about revision cfgs off-topic for this test?

Indeed they are. Added them manually so that no diagnostic is ever emitted.

> Maybe the revision cfgs should be added to known cfgs (possibly automatically by compiletest).

I've taught about this as well, and it's a good idea except for those tests, because having those diagnostics pop up here is very important, because with those tests we want to make sure we're properly testing what we think we're testing, and hiding things will imo only make it more likely to introduce bugs.

[the remark this answers is not in the document]
Comment (petrochenkov) [reply tests/ui/check-cfg/unknown-values.with_values.stderr]: Could you mark all revisition cfgs as expected?
They can be found by "was set with `--cfg` but isn't in the `--check-cfg` expected names".

It's hard to review tests due to so much noise.
The file's revision is already clearly visible from file names, so additional warnings in stderr files are not needed.

[the remark this answers is not in the document]
Comment (Urgau) [reply tests/ui/check-cfg/unknown-values.with_values.stderr]: Sure. I did it for all the tests were it wouldn't have changed the behavior; i.e. except for `exhaustive-names-values.rs`, `exhaustive-names.rs` and `exhaustive-values.rs` where expecting the revisions would be significant. All the other instances are "real" warnings.

@rustbot ready

File: compiler/rustc_interface/src/interface.rs [hunk 3]
```diff
@@ -215,6 +230,116 @@ pub fn parse_check_cfg(handler: &EarlyErrorHandler, specs: Vec<String>) -> Check
...
+                                let mut values_any_specified = false;
+
+                                for arg in args {
+                                    if arg.is_word() && let Some(ident) = arg.ident() {
+                                        if values_specified {
+                                            error!("`cfg()` names cannot be after values");
+                                        }
+                                        names.push(ident);
+                                    } else if let Some(args) = arg.meta_item_list()
+                                        && arg.has_name(sym::any)
```
Comment (petrochenkov) [inline compiler/rustc_interface/src/interface.rs:250]: Could you change the condition order to be like `arg.has_name(sym::any) && let Some(args) = arg.meta_item_list()`?
(Here and in all places below.)

We don't need the list if the argument is not `any`, so this version is easier to read, and also probably faster.

File: tests/ui/check-cfg/unexpected-cfg-name.names.stderr [hunk 4]
```diff
@@ -0,0 +1,10 @@
+warning: unexpected `cfg` condition name
```
Comment (petrochenkov) [inline tests/ui/check-cfg/unexpected-cfg-name.names.stderr:1]:
```suggestion
warning: unexpected `cfg` condition name: widnows
```
This is also an independent change, but could you mention the unexpected names and values in the primary messages as well?
(Could be split to a separate PR together with `should_print_possibilities`.)

**Comment by petrochenkov** [comment]
Reviewed.

File: tests/ui/check-cfg/unexpected-cfg-name.names.stderr (hunk 4 above)
Comment (Urgau) [reply tests/ui/check-cfg/unexpected-cfg-name.names.stderr:1]: Done as part of https://github.com/rust-lang/rust/pull/116666.

**Comment by Urgau** [comment]
The MCP has been [accepted](https://github.com/rust-lang/compiler-team/issues/636#issuecomment-1766619829) :tada:, ready for review/approval.

@rustbot labels -S-waiting-on-MCP -S-waiting-on-team +S-waiting-on-review

**Comment by WaffleLapkin** [comment]
@Urgau is the syntax in the OP still up to date?

**Comment by Urgau** [comment]
> Urgau is the syntax in the OP still up to date? (I don't immediately get what the syntax means, specifically the `values(...)` so I'm asking to make sure)

Yes, the PR description, the commit message and the MCP are all up to date.
Regarding `values(...)` (I assume you mean in `cfg(...)` and not the old syntax), it's a designator to separate the names previously given and the values associated with them. In other words it's like a key=value pair where both the names (key) and values are de-duplicated.

For more information I would highly suggest regarding the [`check-cfg section of the rustc unstable book`](https://github.com/rust-lang/rust/blob/fc78d789887047e128a36bd6e8294e15e57f50a0/src/doc/unstable-book/src/compiler-flags/check-cfg.md#the-cfg-form).

> And is this feature gated? Or is it insta-stable?

It's a pre-existing unstable feature, so yes it's gated.

**Comment by Urgau** [comment]
`values(...)` being it's "own" thing is on purpose, cf. https://hackmd.io/@wesleywiser/rJlumbfep:
> 2. I think we should specifically denote the values. Perhaps `cfg(feature, values(foo, bar))`. No values would be spelled `cfg(feature)` or `cfg(feature, values())`. This makes it easer to extend the constraints we can express in the future (for example `cfg(target_pointer_width, one_of(16, 32, 64))`).

*btw, if you wish to continue discussing it let's do that on Zulip*

----

> Also the documentation seems to have a `configure` [in an example](https://github.com/rust-lang/rust/blob/fc78d789887047e128a36bd6e8294e15e57f50a0/src/doc/unstable-book/src/compiler-flags/check-cfg.md#example-checking-feature-values-but-not-condition-names), which looks like a mistake?

Yes, missed-it.
